course9/lecture_en.md

```diff
@@ -0,0 +1,273 @@
+
+```moonbit
+fn insert[T : Compare](tree : Tree[T], value : T) -> Tree[T] {
+  // Since `T` is bound by `Compare`, it should support the `compare` method.
+  match tree {
+    Empty => Node(value, Empty, Empty) 
+    Node(v, left, right) =>  
+      if T::compare(value, v) == 0 { // We can call `compare` here.
+        tree
+      } else if T::compare(value, v) < 0 { // We can call `compare` here.
+        Node(v, insert(left, value), right)
+      } else {
+        Node(v, left, insert(right, value))
+      }
+  }
+}
+```
+
+## Implementation of Traits
+
+Since the trait system of MoonBit is structured, to implement a trait, all we need to do is to define the corresponding methods using the syntax `fn <type>::<method>(...) -> ...`.
+
+In the following example, we define the `default` method for the `BoxedInt` type.
+
+```moonbit
+struct BoxedInt { value : Int }
+
+fn BoxedInt::default() -> BoxedInt {
+  { value : Int::default() }
+  // The default value can be defined by boxing the default value of `Int`.
+}
+```
+
+By defining the `default` method, the `Default` trait is now implemented. Therefore, we can construct a queue of `BoxedInt` using the `make` function, which internally invokes the `default` method of `BoxedInt`.
+
+```moonbit no-check
+fn init {
+  let array: Queue[BoxedInt] = make()
+}
+```
+
+## Method Chaining
+
+In addition to `<type>::<method>(<expr>, ...)`, we can as well call the method using `<expr>.<method>(...)`, given `<expr>` is of type `<type>`. When defining such methods, `<type>::` can also be omitted when the first parameter is named `self`.
+
+In this way, when we need to call such functions sequentially, we can avoid multiple nesting of method calls, thereby enhancing the readability of the code.
+
+```moonbit
+fn BoxedInt::plus_one(b: BoxedInt) -> BoxedInt {
+  { value : b.value + 1 }
+}
+fn plus_two(self: BoxedInt) -> BoxedInt {
+  // `<type>::` can be omitted when the first parameter is named `self`.
+  { value : self.value + 2}
+}
+
+fn init {
+  let _five = { value: 1 }.plus_one().plus_one().plus_two()
+  // This avoids multiple nesting of method calls.
+  let _five = plus_two(plus_one(plus_one({value: 1})))
+}
+```
+
+## Automatic Derivation of Builtin Traits
+
+In MoonBit, it is possible to automatically derive certain basic traits by adding `derive(<traits>)` after the type declaration. If the type is composed of other types, those member types must have already implemented the same traits.
+
+In the given example, we derive the `Default`, `Eq`, `Compare`, and `Debug` traits for `BoxedInt`. These traits can be successfully derived because `Int` already implements them. However, if the required traits are not implemented by the member types, the compiler will generate an error message.
```

Comment:
TODO: `Debug` will be merged with `Show` some day, so we'll need to update this in the future